--- a/src/control.cc	2018-09-26 09:07:18.646460236 +0100
+++ a/src/control.cc	2018-09-26 09:42:16.138430480 +0100
@@ -168,6 +168,8 @@ Control::is_shutdown_completed() {
 
 void
 Control::handle_shutdown() {
+  rpc::commands.call_catch("event.system.shutdown", rpc::make_target(), "shutdown", "System shutdown event action failed: ");
+
   if (!m_shutdownQuick) {
     // Temporary hack:
     if (worker_thread->is_active())
--- a/src/main.cc	2018-09-10 12:44:24.730779229 +0100
+++ a/src/main.cc	2018-09-26 10:16:56.563826514 +0100
@@ -253,6 +253,9 @@ main(int argc, char** argv) {
        "method.insert = event.view.hide,multi|rlookup|static\n"
        "method.insert = event.view.show,multi|rlookup|static\n"
 
+       "method.insert = event.system.startup_done,multi|rlookup|static\n"
+       "method.insert = event.system.shutdown,multi|rlookup|static\n"
+
        "method.insert = event.download.inserted,multi|rlookup|static\n"
        "method.insert = event.download.inserted_new,multi|rlookup|static\n"
        "method.insert = event.download.inserted_session,multi|rlookup|static\n"
@@ -501,6 +504,8 @@ main(int argc, char** argv) {
 
     worker_thread->start_thread();
 
+    rpc::commands.call_catch("event.system.startup_done", rpc::make_target(), "startup_done", "System startup_done event action failed: ");
+
     torrent::thread_base::event_loop(torrent::main_thread());
 
     control->core()->download_list()->session_save();
