--- src/core/download_factory.cc	2015-09-03 20:03:30.000000000 +0100
+++ src/core/download_factory.cc	2016-07-31 20:49:03.295070681 +0100
@@ -418,6 +418,9 @@ DownloadFactory::initialize_rtorrent(Dow
   if (rtorrent->has_key_value("total_uploaded"))
     download->info()->mutable_up_rate()->set_total(rtorrent->get_key_value("total_uploaded"));
 
+  if (rtorrent->has_key_value("total_downloaded"))
+    download->info()->mutable_down_rate()->set_total(rtorrent->get_key_value("total_downloaded"));
+
   if (rtorrent->has_key_value("chunks_done") && rtorrent->has_key_value("chunks_wanted"))
     download->download()->set_chunks_done(rtorrent->get_key_value("chunks_done"), rtorrent->get_key_value("chunks_wanted"));
 
--- src/core/download_store.cc	2015-08-08 16:40:41.000000000 +0100
+++ src/core/download_store.cc	2016-07-31 20:49:26.544944726 +0100
@@ -140,6 +140,7 @@ DownloadStore::save(Download* d, int fla
   rtorrent_base->insert_key("chunks_done",    d->download()->file_list()->completed_chunks());
   rtorrent_base->insert_key("chunks_wanted",  d->download()->data()->wanted_chunks());
   rtorrent_base->insert_key("total_uploaded", d->info()->up_rate()->total());
+  rtorrent_base->insert_key("total_downloaded", d->info()->down_rate()->total());
 
   // Don't save for completed torrents when we've cleared the uncertain_pieces.
   torrent::resume_save_progress(*d->download(), *resume_base);
