--- rel-0.9.6/src/rpc/command.cc	2015-08-08 16:40:41.000000000 +0100
+++ rtorrent-0.9.6/src/rpc/command.cc	2016-08-24 12:02:46.933616680 +0100
@@ -77,7 +77,7 @@ command_base_call_value_base(command_bas
     return command_base::_call<typename command_value_function<T>::type, T>(rawCommand, target, val);
   }
 
-  return command_base::_call<typename command_value_function<T>::type, T>(rawCommand, target, arg.as_value());
+  return command_base::_call<typename command_value_function<T>::type, T>(rawCommand, target, unit * arg.as_value());
 }
 
 template <typename T> const torrent::Object
