--- src/torrent/tracker_controller.cc	2016-12-12 08:49:18.000000000 +0000
+++ src/torrent/tracker_controller.cc	2018-05-12 18:09:22.922680300 +0100
@@ -391,15 +391,16 @@ tracker_next_timeout_promiscuous(Tracker
       !tracker->is_usable())
     return ~uint32_t();
 
-  int32_t interval;
+  int32_t interval, use_interval;
+  int32_t min_interval = std::max(tracker->min_interval(), (uint32_t)300);
 
-  if (tracker->failed_counter())
+  if (tracker->failed_counter()) {
     interval = 5 << std::min<int>(tracker->failed_counter() - 1, 6);
-  else
+    use_interval = std::min(interval, min_interval);
+  } else {
     interval = tracker->normal_interval();
-
-  int32_t min_interval = std::max(tracker->min_interval(), (uint32_t)300);
-  int32_t use_interval = std::min(interval, min_interval);
+    use_interval = std::max(interval, min_interval);
+  }
 
   int32_t since_last = cachedTime.seconds() - (int32_t)tracker->activity_time_last();
 
